website/src/pages/edited-collections/edited-collection-context.tsx

```diff
@@ -0,0 +1,165 @@
+function findSubchapters(
+  chapters: Chapter[] | undefined,
+  targetLeaf: string
+): Chapter[] | undefined {
+  // If there are no more chapters to be found, return undefined.
+  if (!chapters) {
+    return chapters
+  }
+
+  // Look through all the given chapters.
+  for (let i = 0; i < chapters.length; i++) {
+    let curr = chapters[i]
+
+    // If the current leaf matches the target leaf, then the parent has been found and its subchapters/children can be returned.
+    if (curr?.leaf === targetLeaf) {
+      return curr.children
+    } else {
+      // Else, the target chapter has not been found, so look through all this parent's subchapters/children.
+      const subchapters = findSubchapters(chapters[i]?.children, targetLeaf)
+      // If the target chapter was found, return its subchapters.
+      if (subchapters) {
+        return subchapters
+      }
+    }
+  }
+
+  return undefined
+}
+
+// Chapter type of collection chapters that comes from the backend.
+type FlatChapters =
+  | readonly ({
+      readonly __typename?: "CollectionChapter" | undefined
+    } & Pick<
+      Dailp.CollectionChapter,
+      "section" | "title" | "path" | "indexInParent"
+    >)[]
+  | null
+
+// Converts a flat-list into a nested-list structure.
+function flatToNested(chapters: FlatChapters | undefined) {
+  if (!chapters) {
+    return undefined
+  }
+
+  const nestedChapters: Chapter[] = []
+  const stack: Chapter[] = []
+
+  for (let i = 0; i < chapters.length; i++) {
+    const curr = chapters[i]
+    const leaf = curr?.path[curr?.indexInParent]
+
+    if (curr && leaf) {
+      // Create a new Chapter with the backend chapter's fields.
+      let chapter: Chapter = {
+        title: curr.title,
+        leaf,
+        indexInParent: curr.indexInParent,
+        section: curr.section,
+        children: undefined,
+      }
+
+      // If the index is 1, then this chapter has no parent chapter.
+      if (curr.indexInParent === 1) {
+        // Since this chapter has no parent, it needs to be added to the nested list.
+        nestedChapters.push(chapter)
+        // In case there was a chapter previously, we'll need to pop it off the stack since we now it no longer has any more children to add to it.
+        stack.pop()
+        // Push this chapter onto the stack to check for its children.
+        stack.push(chapter)
+      } else {
+        // Get the item last pushed onto the stack.
+        let lastPushed = stack[stack.length - 1]
+        // Gets the second to last string element in the current chapter's path, which is this chapter's parent leaf.
+        let parentChapterLeaf = curr.path[curr.indexInParent - 1]
+
+        // Check if the current chapter's parent leaf matches the last pushed chapter's leaf.
+        // If it doesn't, the last pushed chapter is not the parent and needs to be popped.
+        // Continue through the stack until the parent of this chapter is found.
+        while (parentChapterLeaf !== lastPushed?.leaf) {
```

Comment:
Such a small thing but is there a way to consolidate the logic of the previous two lines of code with this while loop, where you're checking if chapters in the stack are the parent? If I'm reading correctly it looks like right now you check the first element in the stack, and then have a while loop

Reply:
Oh do you mean like remove the `if (curr.indexInParent === 1) {` and utilize the while loop instead?

Reply:
Actually I just removed most of this code and rewrote it with a `HashMap` instead so it's quicker. If you have time could you look over it again? Tysm!!